libs/editor/egui_editor/src/styled_chunk.rs

```diff
@@ -32,143 +32,154 @@ impl StyledChunk {
     }
 }
 
-impl Editor {
-    /// Traverse an AST and determine 2 key things:
-    /// * The elements that a region of text is impacted by
-    /// * Which regions of text should be in their own galley
-    ///
-    /// This is a recursive function that processes a given node, and it's children. It has to
-    /// capture text that may occur before the first child, between children, and after the last
-    /// child. The rest is handled by recursion.
-    ///
-    /// parent_empty_block serves as a way to signal that the parent is a block, but there was no
-    /// text before we traversed into the first child (think: + *first child* tail).
-    fn region_helper(
-        &mut self, node_idx: usize, parent_elements: &[Element], parent_empty_block: bool,
-    ) {
-        let node = &self.ast.nodes[node_idx];
-        let mut elements = Vec::from(parent_elements);
-        elements.push(node.element.clone());
-
-        let is_block = node.element.is_block() || parent_empty_block;
-
-        if node.children.is_empty() {
-            self.styled.extend(self.cursor_split(StyledChunk {
-                block_start: is_block,
-                range: node.range.clone(),
-                elements,
-            }));
-            return;
-        }
+/// Traverse an AST and determine 2 key things:
+/// * The elements that a region of text is impacted by
+/// * Which regions of text should be in their own galley
+///
+/// This is a recursive function that processes a given node, and it's children. It has to
+/// capture text that may occur before the first child, between children, and after the last
+/// child. The rest is handled by recursion.
+///
+/// parent_empty_block serves as a way to signal that the parent is a block, but there was no
+/// text before we traversed into the first child (think: + *first child* tail).
+pub fn calc(ast: &AST, selection: &Option<Range<DocByteOffset>>) -> Vec<StyleInfo> {
+    let mut styles = Vec::new();
+    calc_recursive(ast, selection, ast.root, &[], false, &mut styles);
+    styles
+}
 
-        let head_range =
-            Range { start: node.range.start, end: self.ast.nodes[node.children[0]].range.start };
+fn calc_recursive(
```

Comment:
`region_helper` renamed to `calc_recursive` and parameters are passed individually instead of using `&mut self`; otherwise unchanged (de-indent/re-wrap make changes look bigger than they are)